Add pattern for discussion: first steps with metrics

dicortazar opened the pull request:
This pattern aims at discussing the process to start
measuring the inner source initiative from a
quantitative point of view.

dicortazar committed:
Add pattern for discussion: first steps with metrics

This pattern aims at discussing the process to start
measuring the inner source initiative from a
quantitative point of view.

gregswindle commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+Metrics are usually misunderstood if people have not received any
+training on those.
+
+
```
> 🙋 Suggested addition

Organizations collect "vanity" metrics that do not track business objectives' success or failure over time.

gruetter replied:
Metrics tend to become goals, will subsequently be gamed and thus meaningless

gregswindle commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+
+Metrics are usually misunderstood if people have not received any
+training on those.
+
```
> 🙋‍♂️  Suggested addition

Organizations have unrealistic (or premature) business objectives. (Needs an example.)

dicortazar replied:
Suggestion added to the text. Does it work for your? Thanks for the suggestion.

gregswindle replied:
LGTM, thanks!

gregswindle commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+# Solution
+
+Bring developers, middle managers and C-level to have a discussion
+about metrics.
```
> 🙋  Suggested edit

Bring developers, middle managers and C-level to have a discussion about business objectives and metrics to track those objectives.

gruetter replied:
Interesting idea, @dicortazar . Do you have experience with this kind of collaboration on metrics by people on different levels in the org chart?

gregswindle replied:
We're working on it, @gruetter. We have not formalized anything, yet, but we plan on:

1. Auto-generating a social graph for "communities of interest" based on Git repository commits by categories defined within [TechRadar Languages & Technologies](https://www.thoughtworks.com/radar/a-z) to segment our internal engineering social network.
1. Using [architecture decision records (ADRs)](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions) to encourage discussions about enterprise standards.
1. Engaging engineering first to generate metrics ADRs about InnerSource metrics.

We're using a "man-on-the-moon" analogy to encourage everyone to thing about metrics as "stages." The metrics collected and reported are different for "launch," "lunar orbit," "lunar landing," etc.

gruetter replied:
I very much like the idea of staging and I can imagine that using ADRs (never heard of these before, thanks for the reference ;) can indeed be helpful as crystallization points for discussions. I'm not sure I got your idea wrt. the _social graph_, though.

gruetter commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+
+The organization does not really know where to start measuring
+things or what are the key parameters to measure.
+
```
Would it make sense to state the goal of the measurements in the problem statement? Perhaps something along the lines of  _The organisation would like to measure the health and performance of the InnerSource initiative but does not really know [...]_. Or maybe _The organisation would like to measure how the InnerSource initiative increases the level of cross-divisional and cross-location collaboration, but does not really know [...]_.

dicortazar replied:
I just added this to the text. Is that how you expected to be?

gruetter commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+
+# Solution
+
+Bring developers, middle managers and C-level to have a discussion
```
There might be other stakeholders, which would want to be part of the discussion. I'm thinking workers councils (might be a german thing). Others, we might to want to include (I'm thinking HR).

gregswindle replied:
@gruetter

I agree. At my employer, we're including

* Developers/engineers
* HR
* IT leadership
* Legal
* Product management
* Sourcing

gruetter replied:
I actually mispoke above: I would suggest to _not include HR_. My fear would be that metrics would quickly "evolve" into personal goals, which would defeat the purpose, IMO.

dicortazar replied:
I've added the verb 'consider' as having HR on board sounds like a necessary way of working. I mean, for instance, should the rewarding system include HR? From my point of view HR should be involved... but other comments?

gruetter commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+Bring developers, middle managers and C-level to have a discussion
+about metrics.
+
+Let developers and middle managers know that these metrics or KPIs
```
Could a solution to the problem of _metrics misuse_ be to have an independent and neutral 3rd party to do the measurement _and interpretation_ of the metrics? (Swiss metrics ;)

gregswindle replied:
@gruetter 

A "neutral third party" might invite bureaucracy, unless its scope is limited to statistical validity. I believe that organizations should create relevant definitions in the following order: 

1. Aim
1. Business objectives
1. Metrics
1. KPIs

I'd love to hear alternative approaches, though!

dicortazar replied:
I'd go for a third party as playing a neutral role in the sense of understanding the needs from C-level, middle management, developers, and others and then implement those in a way understandable by everyone.

This would help when avoiding suspicious behaviours at any level.

@gregswindle having an approach as the one you mention is great. There is a clear need for defining a strategy and method to work on metrics. Indeed I'd go for a couple of extra patterns with this respect that we could reference from here. Does it make sense?

gruetter commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+
+Bring specific discussions on the metrics to be used avoiding per
+developer granularity or at least anonymizing that info.
+
```
Is there good advice on how to start a metrics initiative? Start small/big? Focus/don't focus on impact of InnerSource initiative on bottom line ($)? Use GQM? Should such advice be part of the solution section?

gruetter commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+
+# Resulting Context
+
+The organization builds a monitoring infrastructure where the agreed
```
> agreed metrics

If part of the solution is that all (?) stakeholder, including developers, need to agree on metrics, that should probably go in the solution section, too. Seems important.

gregswindle replied:
@gruetter, I really like the recommendations in ["When, What and How to Measure"](https://dicortazar.gitbooks.io/managing-inner-source-projects/content/measuring/introduction.html) (as you suggested, @dicortazar).

Great work, BTW, @dicortazar!

dicortazar replied:
Thanks @gregswindle !! Does it make sense to elaborate a bit more on the method and strategy? Or does it make sense if we open a new pattern to deal with that specific topic and then link that pattern from here? (I like this second idea... but not sure how to proceed).

gruetter commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,81 @@
+
+Those metrics will help C-level to understand the current situation
+of the project and this will help to compare the status before
+applying inner source and after applying the inner source initiative.
```
wouldn't that require baselining? If yes, should that go in the solution section, as well?

gregswindle replied:
What does everyone think about applying the _same_ metrics to "traditional" internal software development, and using "traditional" development for baselines?

dicortazar replied:
I wouldn't say baselining and that sounds more like historical metrics, but at least have metrics prior and after an event took place.

gruetter reviewed (commented):
I'm done with the first pass review, @dicortazar .

dicortazar committed:
Improve pattern based on reviews: forces and solution

Thanks to @gregswindle, @gruetter and @NewMexicoKid for their pieces of advice
* Update problem statement to mention metrics
* Add forces such as unrealistic objectives, vanity or gaming
* Add other stakeholders to the solution 
* Add third party neutral metrics manager to the solution
* Add specific method and strategy to start with metrics

NewMexicoKid commented on introducing-metrics-in-innersource.md:
```diff
@@ -0,0 +1,98 @@
+# Title
+
+Introducing Metrics in the InnerSource Initiative
```
Change title to: Introducing Metrics In InnerSource

dicortazar committed:
[pattern/metrics] Add last comments from the working session

Thanks to @rrrutledge, @NewMexicoKid, @gruetter, Clint, Bob
and others for the working session.

Modifications were mainly added to the Context and Forces sections.
Changes were focused on stating the early stage of the InnerSource
initiative to start measuring and having a strategy around
metrics.

NewMexicoKid committed:
Merge branch 'master' into pattern/introducing-metrics-in-innersource
